clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java

```diff
@@ -278,18 +303,80 @@
+        else {
+            ConsumerGenerationPair existingPair = prevAssignment.get(partition);
+            if (consumerGeneration.generation > existingPair.generation)
+                prevAssignment.put(partition, consumerGeneration);
+            else if (consumerGeneration.generation == existingPair.generation)
+                // same partition is assigned to two consumers during the same rebalance
+                throw new IllegalStateException("Error: Partition '" + partition + "' had been assigned to " +
+                        "multiple consumers following sticky assignment generation " + consumerGeneration.generation + ".");
+        }
+    }
+
     private void prepopulateCurrentAssignments(Map<String, Subscription> subscriptions,
-                                               Map<String, List<TopicPartition>> currentAssignment) {
+                                               Map<String, List<TopicPartition>> currentAssignment,
+                                               Map<TopicPartition, ConsumerGenerationPair> prevAssignment) {
+        // we need to process subscriptions' user data with each consumer's reported generation in mind
+        // higher generations overwrite lower generations in case of a conflict
+        // note that a conflict could exists only if user data is for different generations
+
+        Map<TopicPartition, ConsumerGenerationPair> partitionConsumer = new HashMap<>();
+        int maxGeneration = 0;
         for (Map.Entry<String, Subscription> subscriptionEntry : subscriptions.entrySet()) {
+            String consumer = subscriptionEntry.getKey();
             ByteBuffer userData = subscriptionEntry.getValue().userData();
-            if (userData != null && userData.hasRemaining())
-                currentAssignment.put(subscriptionEntry.getKey(), deserializeTopicPartitionAssignment(userData));
+            if (userData == null || !userData.hasRemaining()) continue;
+            ConsumerUserData consumerUserData = deserializeTopicPartitionAssignment(userData);
+            if (consumerUserData.generation > maxGeneration)
+                maxGeneration = consumerUserData.generation;
+
+            for (TopicPartition partition: consumerUserData.partitions) {
+                if (!partitionConsumer.containsKey(partition))
+                    partitionConsumer.put(partition, new ConsumerGenerationPair(consumer, consumerUserData.generation));
+                else {
+                    ConsumerGenerationPair existingRecord = partitionConsumer.get(partition);
+                    if (consumerUserData.generation == existingRecord.generation) {
+                        // same partition is assigned to two consumers during the same rebalance
+                        throw new IllegalStateException("Error: Partition '" + partition + "' is assigned to multiple consumers " +
+                                "following sticky assignment generation " + consumerUserData.generation + ".");
```

Comment:
Small remark: By throwing here an exception you are giving up.
But I am wondering if you should not try to fix it.

I do not know which consistency guarantees a kafka consumer group exactly gives.  But if it is possible to get into a split brain situation (network split), where temporarily a consumer group is split in two and gets two leaders.  You get in a situation where the `SitckyAssignor` will never recover.
If it was me, I would write a big fat error, and drop one of the two assignments.

The same for line#322.

Reply:
That makes sense. I updated this to make sure it keeps running.